### PR TITLE
VideoSoftware: Implement xfb copy filter (Deflickering/Brightness)

### PR DESCRIPTION
**Pull request by phire**
The Copy filter does 3 things.

**1)** Resolve the Multisampled framebuffer. Games can specify 7 coefficients, the 3 samples from the current pixel, and two samples from the pixels above and below. 
Since dolphin hasn't implemented Multisampling, even in video software, I haven't implemented this fully.

When games aren't using the multisampling mode, the 3 middle coefficients all point at the current pixel, while the upper/lower coefficients also point at a single upper or lower pixel.

**2)** Deflicker filter. The game does a vertical blur across 3 lines, blending the odd and even fields together which minimises the flickering effect inherent to 480i consoles on interlaced TVs.
This is why 99.9% of games render all 480 lines each frame even when rendering at 60fps.

The filter is counter-productive for progressive displays, which is why dolphin hasn't implemented it before now, though it does blur away the dither pattern. 

When we implement this in the hardware backends, we can add an enhancement which points all 7 coefficients at the middle pixel, forcing the deflicker filter off.

**3)** Brightness filter. You are meant to choose coefficients that add up to 64. But some games cleverly break this rule to create a brightness filter. Resident Evil 4 uses this to implement a global brightness slider option, while other (Rogue Squadron 3?, mario galaxy?) games use it to implement a fade to black effect.

If it wasn't for **3**, there would be little reason to implement this, except so users of melee/brawl could see the deflicker on/off option in action. 

---

This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3845)

### CODE REVIEW

**Comment by lioncash** [comment]
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/VideoCommon/BPMemory.h, line 1027 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3845#-KI-bpo82a6sXze0Vl9v:-KI-bpo82a6sXze0Vl9w:934230794) ([raw file](https://github.com/dolphin-emu/dolphin/blob/4ebed592ad0fa1e339fd756ee1ad0eb870b4c5df/Source/Core/VideoCommon/BPMemory.h#L1027)):*
> ```objc
> 	BitField<44, 6, u64> g;
> 
> 	void getCoefficients(u8 *filterCoefficients)
> ```

`GetCoefficients`

Also, you can just return a std::array.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3845)*

**Comment by JMC47** [comment]
I can't test this.  Software renderer only shows a green screen even after following the instructions.

**Comment by phire** [comment]
Correct wrapping behaviour: 
There is optional clamping at the top and bottom of the copy region (see the UPE_copy struct)

If you don't enable clamping, `0 - 1` will wrap around to 1023, which isn't backed by any memory and returns garbage.

Thanks to Extrems for doing some hardware testing.

**Comment by shuffle2** [comment]
Cool to see it implemented, finally.

File: Source/Core/VideoBackends/Software/EfbInterface.cpp [hunk 1]
```diff
@@ -476,11 +476,35 @@ namespace EfbInterface
...
+			// In non-multisampling mode:
+			//   * Coefficients 2, 3 and 4 sample from the current pixel.
+			//   * Coefficients 0 and 1 sample from the pixel above this one
+			//   * Coefficients 5 and 6 sample from the pixel below this one
+
+			int sum = upperColor[i] * (filterCoefficients[0] + filterCoefficients[1]) +
+			          color[i] * (filterCoefficients[2] + filterCoefficients[3] + filterCoefficients[4]) +
+			          lowerColor[i] * (filterCoefficients[5] + filterCoefficients[6]);
+
+			// TODO: this clamping behavior appears to be correct, but isn't confirmed on hardware.
```
Comment (degasus) [inline Source/Core/VideoBackends/Software/EfbInterface.cpp]: Why? IMO overflow is more likely. Clamping is quite expensive

File: Source/Core/VideoCommon/BPMemory.h [hunk 2]
```diff
@@ -1012,6 +1012,30 @@ union UPE_Copy
...
+
+	BitField< 0, 6, u64> a;
+	BitField< 6, 6, u64> b;
+	BitField<12, 6, u64> c;
+	BitField<18, 6, u64> d;
+	BitField<32, 6, u64> e;
+	BitField<38, 6, u64> f;
+	BitField<44, 6, u64> g;
+
+	void getCoefficients(u8 *filterCoefficients)
```
Comment (degasus) [inline Source/Core/VideoCommon/BPMemory.h]: What do you think about a pointer to u8[7]?

File: Source/Core/VideoBackends/Software/EfbInterface.cpp (hunk 1 above)
Comment (phire) [inline Source/Core/VideoBackends/Software/EfbInterface.cpp]: I wasn't expecting clamping, but Resident evil 4's glitches if you push the brightness slider too high without clamping.

Besides, a clamp at 255 can be implemented with 8 muxes hooked up to the carry out of the adder.

**Comment by sepalani** [comment]
Is there a reason for ```const std::array<u8, 7> filterCoefficients``` not being const-reference?

**Comment by lioncash** [comment]
Review status: 0 of 5 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[Source/Core/VideoBackends/Software/EfbInterface.cpp, line 554 \[r3\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3845#-KI6XoXvVqWn21-Grl0S:-KI6XoXwVmHMIgTBPwch:-56593956) ([raw file](https://github.com/dolphin-emu/dolphin/blob/97e7694066cc98b6ea6cb5cc8dc1607b68434585/Source/Core/VideoBackends/Software/EfbInterface.cpp#L554)):*
> ```cpp
> 			//
> 			//         In our implementation, the garbage just so happens to be the top or bottom row.
> 			//         Statically, that could happen.
> ```

Should this be 'Statistically'?

---

*[Source/Core/VideoCommon/BPMemory.h, line 1027 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3845#-KI-bpo82a6sXze0Vl9v:-KI6WPcMRpFXPL_n2Dk_:428836363) ([raw file](https://github.com/dolphin-emu/dolphin/blob/4ebed592ad0fa1e339fd756ee1ad0eb870b4c5df/Source/Core/VideoCommon/BPMemory.h#L1027)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> `GetCoefficients`
> 
> Also, you can just return a std::array.
</details>
Still needs to be `GetCoefficients()` (also it can be made const qualified as well).

---

*[Source/Core/VideoCommon/BPMemory.h, line 1027 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3845#-KI-XgsVDy5DI0AtRYRH-r1-1027:-KI6XCyqFY89OZ9I2CZO:1442009774) ([raw file](https://github.com/dolphin-emu/dolphin/blob/4ebed592ad0fa1e339fd756ee1ad0eb870b4c5df/Source/Core/VideoCommon/BPMemory.h#L1027)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> What do you think about a pointer to u8[7], or better, u8[7] as return value.
</details>
A std::array is better in this situation, IMO.

---

*[Source/Core/VideoCommon/BPMemory.h, line 1039 \[r3\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3845#-KI6YfIsJE2N8Kr8zoUV:-KI6YfIsJE2N8Kr8zoUW:632198586) ([raw file](https://github.com/dolphin-emu/dolphin/blob/97e7694066cc98b6ea6cb5cc8dc1607b68434585/Source/Core/VideoCommon/BPMemory.h#L1039)):*
> ```objc
> 		};
> 	}
> ```

unnecessary newline?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3845)*

**Comment by degasus** [comment]
What do you think about std:array<u8, 8>? So it will use exactly one register:  https://godbolt.org/g/TbA0LM

**Comment by lioncash** [comment]
@degasus I'm not particularly sure what you're asking, but if you mean modify GetCoefficients so it returns 8 elements instead of 7, then I can only really say "please don't micro-optimize like that", by adding completely unnecessary elements to the array. GetCoefficients means 'get the coefficients' not 'get the coefficients and a dummy entry'. That also makes it look like a potential off-by-one bug to anyone not familiar.

That test isn't particularly indicative of a gain either, since it's literally evaluating the induction variable results and eliding the loop entirely (since it knows the fixed constants at that point). It's also using a looping context where it wouldn't be used to begin with.

You'd typically get something like [this](https://godbolt.org/g/WyGfMU) instead. We currently compile under O2 as well, so even then, we'd still not get that level of optimization to begin with.

**Comment by phire** [comment]
I consolted godbolt and when you [make the code more complex](https://godbolt.org/g/unqTOl), `std::array<u8, 8>` actually compiles worse on gcc (but identical on clang).

However, passing by constant reference is faster and produces identical code for both on both compilers.

**Comment by phire** [comment]
For refrence, here is a version that uses our BitField class: https://godbolt.org/g/LAj294

GCC doesn't optimise the unpacking as much.

**Comment by Fallcrest** [comment]
@phire When you mention Mario Galaxy, does this PR fix the missing roar effects? ([Issue 8327](https://bugs.dolphin-emu.org/issues/8327))

**Comment by phire** [comment]
It was theorized by @JMC47 that it might at the time this commit message was written. However later testing shows otherwise.

**Comment by Fallcrest** [comment]
@phire So there is no current explanation for the missing effect? It is so odd that something that minor has escaped finding for so long. Are there any other effects missing from other games that might be related?

**Comment by JMC47** [comment]
Can this be rebased and merged?  Would be nice to update the current oldest issue in dolphin :D

**Comment by RisingFog** [comment]
This PR still need to be rebased.

**Comment by Buddybenj** [comment]
Any update on this?

**Comment by Buddybenj** [comment]
Is there any chance this will be rebased?

**Pull request closed without merging**